modules/MarketAnalysis.py

```diff
@@ -25,79 +41,131 @@ def __init__(self, config, api):
         self.api = api
         self.lending_style = int(config.get('BOT', 'lendingStyle', 50, 1, 99))
+        self.recorded_levels = 90
+        self.modules_dir = os.path.dirname(os.path.realpath(__file__))
+        self.top_dir = os.path.dirname(self.modules_dir)
+        self.db_dir = os.path.join(self.top_dir, 'market_data')
 
         if len(self.currencies_to_analyse) != 0:
             for currency in self.currencies_to_analyse:
-
                 try:
                     self.api.api_query("returnLoanOrders", {'currency': currency, 'limit': '5'})
                 except Exception as cur_ex:
                     print "Error: You entered an incorrect currency: '" + currency + \
                           "' to analyse the market of, please check your settings. Error message: " + str(cur_ex)
                     exit(1)
 
-                else:
-                    path = "market_data/" + currency + "_market_data.csv"
-                    self.open_files[currency] = path
-
     def run(self):
-        thread = threading.Thread(target=self.update_market_loop)
+        for cur in self.currencies_to_analyse:
+            db_con = self.create_connection(cur)
+            self.create_rate_table(db_con, self.recorded_levels)
+            db_con.close()
+        thread = threading.Thread(target=self.run_threads)
         thread.deamon = True
         thread.start()
+        # del_thread = threading.Thread(target=self.run_del_threads)
+        # del_thread.deamon = True
+        # del_thread.start()
 
-    def update_market_loop(self):
+    def run_threads(self):
+        while True:
+            for cur in self.currencies_to_analyse:
+                thread = threading.Thread(target=self.update_market_thread, args=(cur,))
+                thread.deamon = False
+                thread.start()
+            # TODO Set this back to the config value
+            time.sleep(1)
+
+    def run_del_threads(self):
+        while True:
+            for cur in self.currencies_to_analyse:
+                del_thread = threading.Thread(target=self.delete_old_data_thread, args=(cur,))
+                del_thread.daemon = True
+                del_thread.start()
+            # TODO set a reasonable default and allow config
+            time.sleep(30)
+
+    def update_market_thread(self, cur):
+        try:
+            db_con = self.create_connection(cur)
+            self.update_market(db_con, cur, self.recorded_levels)
+        except Exception as ex:
+            ex.message = ex.message if ex.message else str(ex)
+            print("Error in MarketAnalysis: {0}".format(ex.message))
+            traceback.print_exc()
+
+    def delete_old_data_thread(self, cur):
+        time_in_sec = 1800  # TODO Take this from config
         while True:
             try:
-                self.update_markets()
-                self.delete_old_data()
+                db_con = self.create_connection(cur)
```

Comment:
It's the connection to the sqlite database created on disk. In this implementation there's a db per currency as I was trying to increase concurrency. I'm still not sure that's the best way to do it though as I'm still having issues deleting and sometimes (though less frequently) on writes.